File: mantleheatflux.m
function [qm,Db,uc,Ra,nu] = mantleheatflux(Tm,Ts,rs,Rp,Rc,g,rho_m,FH2O,meltfrac)
%This function calculates the mantle heat flux, which depends on the
%temperature and viscosity of the mantle. The viscosity depends on
%temperature and is calculated with the function viscosity.m

%mantle thermal conductivity (W/m/K)
km = 4.2;
%depth of the convective zone
% if rs < Rp
if meltfrac >=0.4
    Z = Rp - rs; %convection only occurs within the magma ocean until solidification
else
    Z = Rp - Rc;
end

% coefficient of thermal expansion (1/K)
alpha = 2e-5;
%mantle heat capacity (J/kg/K)
cp = 1.2e3; 
%mantle thermal diffusivity (m^2 / s)
kappa = km / rho_m / cp;

%find the viscosity (m^2/s)
% if Ts < 1420
%     nu = viscosity2(Tm,FH2O,g,rho_m,0);
% else
    nu = viscosity(Tm,Ts,rho_m, FH2O, meltfrac);
% end

%Calculate Rayleigh number
Ra = (g * alpha * abs(Tm - Ts) * Z^3) / nu / kappa;

% %boundary layer depth (m)
% Db = Z * (Ra_cr / Ra)^beta;
qm = 0.089 * km * (Tm - Ts) * Ra^(1/3) / Z;
Db = km * (Tm - Ts) / qm;
    
%spreading time (s)
ts = Db^2 / (5.38 * kappa);%s

%spreading velocity (m/s)
uc = Z/ts; %m/s

end

File: physics/mantleheatflux.py
from .viscosity import viscosity

def mantleheatflux(
        temp_mantle,
        temp_surface,
        radius_solid,
        radius_planet,
        radius_core,
        gravity,
        density_mantle,
        mass_frac_water,
        melt_fraction,
        params):
    """
    Calculates the convective heat flux and boundary layer properties of the mantle 
    using parameterized Rayleigh-Bénard convection scaling.

    The convective depth is set by the caller via radius_solid; this function
    does not determine the phase internally.

    Parameters
    ----------
    temp_mantle : float
        Potential temperature of the convective mantle (K).
    temp_surface : float
        Surface temperature of the planet (K).
    radius_solid : float
        Base of the active convective zone (m). The caller sets this to the
        solidification front during the MO phase, or to radius_core in the
        solid-mantle phase. radius_core is accepted but unused.
    radius_planet : float
        Total radius of the planet (m).
    radius_core : float
        Unused; kept for call-site compatibility.
    gravity : float
        Surface gravitational acceleration (m/s^2).
    density_mantle : float
        Bulk density of the mantle (kg/m^3).
    mass_frac_water : float
        Mass fraction of water in the convecting region (dimensionless).
    melt_fraction : float
        Volume-averaged melt fraction of the convecting region (dimensionless).
    params : dict
        Main configuration dictionary containing TOML parameters.

    Returns
    -------
    heat_flux_mantle : float
        The convective heat flux extracted from the mantle (W/m^2).
    depth_boundary_layer : float
        The thickness of the upper thermal boundary layer (m).
    velocity_spreading : float
        The characteristic convective velocity / surface spreading rate (m/s).
    rayleigh_number : float
        The dimensionless Rayleigh number of the convective system.
    kinematic_viscosity : float
        The kinematic viscosity of the convecting material (m^2/s).
    """

    # --- Unpack Parameters ---
    thermo = params['planet']['thermodynamics']
    conv   = params['planet']['convection']

    # --- Thermodynamic Constants ---
    thermal_conductivity = thermo['thermal_conductivity'] # km (W/m/K)
    thermal_expansion    = thermo['thermal_expansion']    # alpha (1/K)
    heat_capacity        = thermo['specific_heat_mantle'] # cp (J/kg/K)

    # --- Convective Geometry ---
    # The caller sets radius_solid to the base of the active convective zone:
    # the solidification front during the MO phase, or radius_core in the solid phase.
    depth_convective_zone = radius_planet - radius_solid

    # --- Fluid Dynamics Properties ---
    # Thermal diffusivity (m^2/s)
    thermal_diffusivity = thermal_conductivity / (density_mantle * heat_capacity)
    
    # Kinematic viscosity (m^2/s)
    kinematic_viscosity = viscosity(temp_mantle, temp_surface, density_mantle, mass_frac_water, melt_fraction, params)

    # --- Rayleigh Number Calculation ---
    temp_difference = abs(temp_mantle - temp_surface)
    
    rayleigh_number = (gravity * thermal_expansion * temp_difference * depth_convective_zone**3) / (kinematic_viscosity * thermal_diffusivity)

    # --- Heat Flux & Boundary Layer Parameters ---
    # Convective heat flux scaling for hard-turbulent regime (Ra^1/3)
    nusselt_coefficient = conv['nusselt_coefficient']
    heat_flux_mantle = nusselt_coefficient * thermal_conductivity * temp_difference * (rayleigh_number**(1.0 / 3.0)) / depth_convective_zone

    # Thermal boundary layer thickness via Fourier's Law of Conduction (m)
    depth_boundary_layer = thermal_conductivity * temp_difference / heat_flux_mantle

    # Characteristic convective spreading time (s)
    spreading_factor = conv['spreading_time_factor']
    time_spreading = (depth_boundary_layer**2) / (spreading_factor * thermal_diffusivity)

    # Characteristic surface spreading velocity (m/s)
    velocity_spreading = depth_convective_zone / time_spreading

    return heat_flux_mantle, depth_boundary_layer, velocity_spreading, rayleigh_number, kinematic_viscosity
